File: spriteforest-ios/action/SFAction.m
//
//  SFAction.m
//  SpriteForest2
//
//  Created by Moky on 12-10-1.
//  Copyright 2012 Slanissue.com. All rights reserved.
//

#import "SFAction.h"

#import "SFActionInterval.h"
#import "SFActionMagic.h"
#import "SFNode.h"


@implementation SFAction

+ (id) getInstance:(NSDictionary *)dict
{
	NSString * name = [dict objectForKey:@"Name"];
	if (![name isKindOfClass:[NSString class]])
	{
		CNLog(@"ERROR: dict = %@", dict);
		return nil;
	}
	Class class = NSClassFromString([@"SF" stringByAppendingString:name]);
	if (![class isSubclassOfClass:[SFAction class]])
	{
		class = [SFActionMagic class];
	}
	return [[[class alloc] initWithDictionary:dict] autorelease];
}

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [self init]))
	{
		//CNLog(@"class = %@, dict = %@", [self class], dict);
	}
	return self;
}

- (CCAction *) ccAction
{
	CNLog(@"override me");
	return nil;
//	return [CCAction action];
}

+ (CCAction *) actionByReplaceTarget:(CCNode *)node dictionary:(NSDictionary *)dict
{
	id target = nil;
	if ([node isKindOfClass:[CCNode class]])
	{
		target = [node getTarget:[dict objectForKey:@"Target"]];
	}
	
	if ([target isKindOfClass:[CCNode class]])
	{
		dict = [NSMutableDictionary dictionaryWithDictionary:dict];
		[(NSMutableDictionary *)dict setObject:target forKey:@"Target"];
	}
	
	return [[self getInstance:dict] ccAction];
}

+ (CCAction *) subActionFromDictionary:(NSDictionary *)dict
{
	id target = [dict objectForKey:@"Target"];
	id action = [dict objectForKey:@"Action"];
	if ([action isKindOfClass:[NSDictionary class]])
	{
		action = [self actionByReplaceTarget:target dictionary:action];
	}
	
	if ([action isKindOfClass:[CCAction class]])
	{
		return action;
	}
	else
	{
		CNLog(@"ERROR!");
		return nil;
	}
}

+ (NSArray *) subActionsFromDictionary:(NSDictionary *)dict
{
	id target = [dict objectForKey:@"Target"];
	NSArray * actions = [dict objectForKey:@"Actions"];
	if (![actions isKindOfClass:[NSArray class]])
	{
		return nil;
	}
	
	NSMutableArray * mArray = [NSMutableArray arrayWithCapacity:[actions count]];
	for (id action in actions)
	{
		NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
		if ([action isKindOfClass:[NSDictionary class]])
		{
			action = [self actionByReplaceTarget:target dictionary:action];
		}
		
		if ([action isKindOfClass:[CCAction class]])
		{
			[mArray addObject:action];
		}
		else
		{
			CNLog(@"ERROR!");
		}
		[pool release];
	}
	
	return mArray;
}

@end


#pragma mark -


@implementation SFFiniteTimeAction

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		duration_ = [[dict objectForKey:@"Duration"] floatValue];
	}
	return self;
}

- (CCAction *) ccAction
{
	CNLog(@"override me");
	return nil;
//	return [CCFiniteTimeAction action];
}

@end


@implementation SFRepeatForever

@synthesize action = action_;

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		id action = [[self class] subActionFromDictionary:dict];
		if ([action isKindOfClass:[CCActionInterval class]])
		{
			self.action = action;
		}
	}
	return self;
}

- (id) init
{
	if ((self = [super init]))
	{
		self.action = nil;
	}
	return self;
}

- (void) dealloc
{
	[action_ release];
	[super dealloc];
}

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: actioin = %@", action_);
		return [super ccAction];
	}
	return [CCRepeatForever actionWithAction:action_];
}

@end


@implementation SFSpeed

@synthesize action = action_;

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		// drived action
		id action = [[self class] subActionFromDictionary:dict];
		if ([action isKindOfClass:[CCActionInterval class]])
		{
			self.action = action;
		}
		
		// speed rate
		rate_ = [[dict objectForKey:@"Rate"] intValue];
	}
	return self;
}

- (id) init
{
	if ((self = [super init]))
	{
		self.action = nil;
	}
	return self;
}

- (void) dealloc
{
	[action_ release];
	[super dealloc];
}

- (CCAction *) ccAction
{
	if (![action_ isKindOfClass:[CCActionInterval class]])
	{
		CNLog(@"Error: actioin = %@", action_);
		return [super ccAction];
	}
	return [CCSpeed actionWithAction:action_ speed:rate_];
}

@end


@implementation SFFollow

@synthesize target = target_;
@synthesize followedNode = followedNode_;
@synthesize rect = rect_;

- (id) initWithDictionary:(NSDictionary *)dict
{
	if ((self = [super initWithDictionary:dict]))
	{
		self.target = [dict objectForKey:@"Target"];
		self.followedNode = [dict objectForKey:@"FollowedNode"];
		
		NSString * rect = [dict objectForKey:@"Rect"];
		if (rect)
		{
			self.rect = CGRectFromString(rect);
		}
	}
	return self;
}

- (id) init
{
	if ((self = [super init]))
	{
		self.target = nil;
		self.followedNode = nil;
		self.rect = CGRectZero;
	}
	return self;
}

- (void) dealloc
{
	[target_ release];
	[followedNode_ release];
	[super dealloc];
}

- (CCAction *) ccAction
{
	if (![target_ isKindOfClass:[CCNode class]])
	{
		CNLog(@"invalid target: %@", target_);
		return [super ccAction];
	}
	if ([followedNode_ isKindOfClass:[NSString class]])
	{
		followedNode_ = [target_ getTarget:followedNode_];
	}
	if (![followedNode_ isKindOfClass:[CCNode class]])
	{
		CNLog(@"Error: followedNode = %@", followedNode_);
		return [super ccAction];
	}
	
	if (CGRectEqualToRect(rect_, CGRectZero))
	{
		return [CCFollow actionWithTarget:followedNode_];
	}
	else
	{
		return [CCFollow actionWithTarget:followedNode_ worldBoundary:rect_];
	}
}

@end

File: x/action/SFAction.cpp
//
//  SFAction.cpp
//  SpriteForest-X
//
//  Created by Moky on 12-11-26.
//  Copyright 2012 Slanissue Inc. All rights reserved.
//

#include "SFAction.h"

#include "../base/SFNode.h"
#include "SFActionMagic.h"

NS_SF_BEGIN

USING_NS_CC;
USING_NS_CN;

#pragma mark SFAction

SFAction::SFAction(void)
: CCObject()
{
	
}

SFAction::~SFAction(void)
{
//	CNLog("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ dealloc");
}

bool SFAction::init(void)
{
	return true;
}

SFAction * SFAction::getInstance(CCDictionary * pDict)
{
	CCString * pName = (CCString *)pDict->objectForKey("Name");
	if (!pName)
	{
		CNLog("error");
		return NULL;
	}
	
	SFAction * pAction = NULL;
	
	std::string sClassName = std::string("SFAction").append(pName->getCString());
	CCObject * pObject = CNRuntimeClass::createObject(sClassName.c_str());
	if ((pAction = dynamic_cast<SFAction *>(pObject)))
	{
		return pAction->initWithDictionary(pDict) ? pAction : NULL;
	}
	
	sClassName = std::string("SF").append(pName->getCString());
	pObject = CNRuntimeClass::createObject(sClassName.c_str());
	if ((pAction = dynamic_cast<SFAction *>(pObject)))
	{
		return pAction->initWithDictionary(pDict) ? pAction : NULL;
	}
	
	sClassName = std::string("SFActionMagic");
	pObject = CNRuntimeClass::createObject(sClassName.c_str());
	if ((pAction = dynamic_cast<SFAction *>(pObject)))
	{
		return pAction->initWithDictionary(pDict) ? pAction : NULL;
	}
	
	CNLog("no such action: %s", pName->getCString());
	return NULL;
}

bool SFAction::initWithDictionary(CCDictionary * pDict)
{
	if (!init())
	{
		CNLog("error");
		return false;
	}
	if (!pDict)
	{
		CNLog("dict cannot be null");
		return false;
	}
	return true;
}

CCAction * SFAction::ccAction(void)
{
	CNLog("override me!");
	return NULL;
}

CCAction * CCAction_actionByReplaceTarget(CCObject * pTarget, CCDictionary * pDict)
{
	if (!pDict)
	{
		CNLog("could not happen!");
		return NULL;
	}
	
	if (!pTarget)
	{
		CNLog("target is NULL!");
		return NULL;
	}
	else if (CCNode * pNode = dynamic_cast<CCNode *>(pTarget))
	{
		pTarget = pDict->objectForKey("Target");
		if (!pTarget)
		{
			pTarget = pDict->objectForKey("ss_Target");
		}
		if (!pTarget)
		{
			pDict->setObject(pNode, "Target");
		}
		else if (CCString * pString = dynamic_cast<CCString *>(pTarget))
		{
			pNode = SFNode::getTarget(pNode, pString->getCString());
			if (pNode)
			{
				pDict->setObject(pNode, "Target");
			}
		}
	}
	else if (CCString * pString = dynamic_cast<CCString *>(pTarget))
	{
		CNLog("error: %s", pString->getCString());
		CC_UNUSED_PARAM(pString);
		return NULL;
	}
	else
	{
		CNLog("could not happen");
		return NULL;
	}
	
	SFAction * pAction = SFAction::getInstance(pDict);
	if (!pAction)
	{
		return NULL;
	}
	return pAction->ccAction();
}

CCAction * SFAction::subActionFromDictionary(CCDictionary * pDict)
{
	if (!pDict)
	{
		CNLog("could not happen!");
		return NULL;
	}
	CCObject * pAction = pDict->objectForKey("Action");
	if (!pAction)
	{
		CNLog("error");
		return NULL;
	}
	
	if (CCDictionary * dict = dynamic_cast<CCDictionary *>(pAction)) // maybe already replaced
	{
		CCObject * pTarget = pDict->objectForKey("Target");
		if (!pTarget)
		{
			pTarget = pDict->objectForKey("ss_Target");
		}
		CCAssert(pTarget, "target not set");
		pAction = CCAction_actionByReplaceTarget(pTarget, dict);
	}
	
	return dynamic_cast<CCAction *>(pAction);
}

CCArray * SFAction::subActionsFromDictionary(CCDictionary * pDict)
{
	if (!pDict)
	{
		CNLog("could not happen!");
		return NULL;
	}
	CCArray * pActions = CNDictionary::arrayForKey("Actions", pDict);
	if (!pActions)
	{
		CNLog("could not happen!");
		return NULL;
	}
	
	CCObject * pTarget = pDict->objectForKey("Target");
	if (!pTarget)
	{
		pTarget = pDict->objectForKey("ss_Target");
	}
	CCAssert(pTarget, "target not set");
	CCArray * pArr = CCArray::arrayWithCapacity(pActions->count());
	
	CCPoolManager * pm = CCPoolManager::sharedPoolManager();
	CCObject * pObj = NULL;
	CCARRAY_FOREACH(pActions, pObj)
	{
		pm->push();
		
		if (CCDictionary * dict = dynamic_cast<CCDictionary *>(pObj)) // maybe already replaced
		{
			pObj = CCAction_actionByReplaceTarget(pTarget, dict);
		}
		
		if (dynamic_cast<CCAction *>(pObj))
		{
			pArr->addObject(pObj);
		}
		else
		{
			CNLog("error");
		}
		
		pm->pop();
	}
	return pArr;
}

#pragma mark SFFiniteTimeAction

SFFiniteTimeAction::SFFiniteTimeAction(void)
: SFAction()
, m_fDuration(0)
{
	
}

bool SFFiniteTimeAction::init(void)
{
	m_fDuration = 0.0f;
	return true;
}

bool SFFiniteTimeAction::initWithDictionary(CCDictionary * pDict)
{
	if (!SFAction::initWithDictionary(pDict))
	{
		CNLog("error");
		return false;
	}
	CCString * pDuration = (CCString *)pDict->objectForKey("Duration");
	if (pDuration)
	{
		m_fDuration = pDuration->floatValue();
	}
	return true;
}

#pragma mark SFRepeatForever

SFRepeatForever::SFRepeatForever(void)
: SFAction()
, m_pAction(NULL)
{
	
}

SFRepeatForever::~SFRepeatForever(void)
{
	setAction(NULL);
}

bool SFRepeatForever::init(void)
{
	setAction(NULL);
	return true;
}

bool SFRepeatForever::initWithDictionary(CCDictionary * pDict)
{
	if (!SFAction::initWithDictionary(pDict))
	{
		CNLog("error");
		return false;
	}
	CCAction * pAction = SFAction::subActionFromDictionary(pDict);
	CCActionInterval * pActionInterval = dynamic_cast<CCActionInterval *>(pAction);
	if (!pActionInterval)
	{
		CNLog("error");
		return false;
	}
	setAction(pActionInterval);
	return true;
}

CCAction * SFRepeatForever::ccAction(void)
{
	if (!m_pAction)
	{
		CNLog("error");
		return NULL;
	}
	return CCRepeatForever::actionWithAction(m_pAction);
}

#pragma mark SFSpeed

SFSpeed::SFSpeed(void)
: SFAction()
, m_pAction(NULL)
, m_fRate(0)
{
	
}

SFSpeed::~SFSpeed(void)
{
	setAction(NULL);
	setRate(0);
}

bool SFSpeed::init(void)
{
	setAction(NULL);
	setRate(0);
	return true;
}

bool SFSpeed::initWithDictionary(CCDictionary * pDict)
{
	if (!SFAction::initWithDictionary(pDict))
	{
		CNLog("error");
		return false;
	}
	CCAction * pAction = SFAction::subActionFromDictionary(pDict);
	CCActionInterval * pActionInterval = dynamic_cast<CCActionInterval *>(pAction);
	if (!pActionInterval)
	{
		CNLog("error");
		return NULL;
	}
	setAction(pActionInterval);
	
	CCString * pRate = (CCString *)pDict->objectForKey("Rate");
	if (pRate)
	{
		m_fRate = pRate->floatValue();
	}
	
	return true;
}

CCAction * SFSpeed::ccAction(void)
{
	if (!m_pAction)
	{
		CNLog("error");
		return NULL;
	}
	return CCSpeed::actionWithAction(m_pAction, m_fRate);
}

#pragma mark SFFollow

SFFollow::SFFollow(void)
: SFAction()
, m_pTarget(NULL)
, m_pFollowedNode(NULL)
, m_tRect(CCRectZero)
{
	
}

SFFollow::~SFFollow(void)
{
	setTarget(NULL);
	setFollowedNode(NULL);
}

bool SFFollow::init(void)
{
	setTarget(NULL);
	setFollowedNode(NULL);
	setRect(CCRectZero);
	return true;
}

bool SFFollow::initWithDictionary(CCDictionary * pDict)
{
	if (!SFAction::initWithDictionary(pDict))
	{
		CNLog("error");
		return false;
	}
	CCObject * pTarget = pDict->objectForKey("Target");
	if (!pTarget)
	{
		pTarget = pDict->objectForKey("ss_Target");
	}
	setTarget(pTarget);
	setFollowedNode(pDict->objectForKey("FollowedNode"));
	
	CCString * pRect = (CCString *)pDict->objectForKey("Rect");
	if (pRect)
	{
		setRect(CCRectFromString(pRect->getCString()));
	}
	return true;
}

CCAction * SFFollow::ccAction(void)
{
	// check followed node
	if (CCString * pString = dynamic_cast<CCString *>(m_pFollowedNode))
	{
		if (CCNode * pNode = dynamic_cast<CCNode *>(m_pTarget))
		{
			m_pFollowedNode = SFNode::getTarget(pNode, pString->getCString());
		}
	}
	
	CCNode * pNode = dynamic_cast<CCNode *>(m_pFollowedNode);
	if (!pNode)
	{
		CNLog("invalid followed node");
		return NULL;
	}
	
	if (CCRect::CCRectEqualToRect(m_tRect, CCRectZero))
	{
		return CCFollow::actionWithTarget(pNode);
	}
	else
	{
		return CCFollow::actionWithTarget(pNode, m_tRect);
	}
}

NS_SF_END
